removing `skipLibCheck` causes type errors (even on the starter template created by create-tsrouter)

### Which project does this relate to?

Router

### Describe the bug

The title is self explanatory, removing `skipLibCheck` from tsconfig causes type errors from the `tanstack/router-core` pacakge. I can't have `skipLibCheck` because I want my d.ts files to be checked.

```
node_modules/.pnpm/@tanstack+router-core@1.119.0/node_modules/@tanstack/router-core/dist/esm/route.d.ts:370:22 - error TS2420: Class 'BaseRoute<TParentRoute, TPath, TFullPath, TCustomId, TId, TSearchValidator, TParams, TRouterContext, ... 5 more ..., TFileRouteTypes>' incorrectly implements interface 'Route<TParentRoute, TPath, TFullPath, TCustomId, TId, TSearchValidator, TParams, TRouterContext, ... 5 more ..., TFileRouteTypes>'.
  Type 'BaseRoute<TParentRoute, TPath, TFullPath, TCustomId, TId, TSearchValidator, TParams, TRouterContext, ... 5 more ..., TFileRouteTypes>' is missing the following properties from type 'Route<TParentRoute, TPath, TFullPath, TCustomId, TId, TSearchValidator, TParams, TRouterContext, ... 5 more ..., TFileRouteTypes>': useMatch, useRouteContext, useSearch, useParams, and 4 more.

370 export declare class BaseRoute<in out TParentRoute extends AnyRoute = AnyRoute, in out TPath extends string = '/', in out TFullPath extends string = ResolveFullPath<TParentRoute, TPath>, in out TCustomId extends string = string, in out TId extends string = ResolveId<TParentRoute, TCustomId, TPath>, in out TSearchValidator = undefined, in out TParams = ResolveParams<TPath>, in out TRouterContext = AnyContext, in out TRouteContextFn = AnyContext, in out TBeforeLoadFn = AnyContext, in out TLoaderDeps extends Record<string, any> = {}, in out TLoaderFn = undefined, in out TChildren = unknown, in out TFileRouteTypes = unknown> implements Route<TParentRoute, TPath, TFullPath, TCustomId, TId, TSearchValidator, TParams, TRouterContext, TRouteContextFn, TBeforeLoadFn, TLoaderDeps, TLoaderFn, TChildren, TFileRouteTypes> {
                         ~~~~~~~~~

node_modules/.pnpm/@tanstack+router-core@1.119.0/node_modules/@tanstack/router-core/dist/esm/route.d.ts:403:23 - error TS2344: Type 'this' does not satisfy the constraint 'AnyRoute'.
  Type 'BaseRoute<TParentRoute, TPath, TFullPath, TCustomId, TId, TSearchValidator, TParams, TRouterContext, ... 5 more ..., TFileRouteTypes>' is missing the following properties from type 'Route<any, any, any, any, any, any, any, any, any, any, any, any, any, any>': useMatch, useRouteContext, useSearch, useParams, and 4 more.

403     lazy: RouteLazyFn<this>;
                          ~~~~
```

### Your Example Website or App

-

### Steps to Reproduce the Bug or Issue

1. npx create-tsrouter-app@latest my-app --template file-router
2. remove `skipLibCheck` from tsconfig
3. run `tsc --build --noEmit`

### Expected behavior

I expect no type errors to occure.

### Screenshots or Videos

_No response_

### Platform

- OS: [Linux]
- Browser: not relevant
- Version: [1.120.2]


### Additional context

_No response_

--- a/packages/react-router/src/route.tsx
+++ b/packages/react-router/src/route.tsx
@@ -25,9 +25,11 @@ import type {
   ResolveFullPath,
   ResolveId,
   ResolveParams,
+  RootRoute as RootRouteCore,
   RootRouteId,
   RootRouteOptions,
   RouteConstraints,
+  Route as RouteCore,
   RouteIds,
   RouteMask,
   RouteOptions,
@@ -148,43 +150,62 @@ export class RouteApi<
 }
 
 export class Route<
-  in out TParentRoute extends RouteConstraints['TParentRoute'] = AnyRoute,
-  in out TPath extends RouteConstraints['TPath'] = '/',
-  in out TFullPath extends RouteConstraints['TFullPath'] = ResolveFullPath<
+    in out TParentRoute extends RouteConstraints['TParentRoute'] = AnyRoute,
+    in out TPath extends RouteConstraints['TPath'] = '/',
+    in out TFullPath extends RouteConstraints['TFullPath'] = ResolveFullPath<
+      TParentRoute,
+      TPath
+    >,
+    in out TCustomId extends RouteConstraints['TCustomId'] = string,
+    in out TId extends RouteConstraints['TId'] = ResolveId<
+      TParentRoute,
+      TCustomId,
+      TPath
+    >,
+    in out TSearchValidator = undefined,
+    in out TParams = ResolveParams<TPath>,
+    in out TRouterContext = AnyContext,
+    in out TRouteContextFn = AnyContext,
+    in out TBeforeLoadFn = AnyContext,
+    in out TLoaderDeps extends Record<string, any> = {},
+    in out TLoaderFn = undefined,
+    in out TChildren = unknown,
+    in out TFileRouteTypes = unknown,
+  >
+  extends BaseRoute<
     TParentRoute,
-    TPath
-  >,
-  in out TCustomId extends RouteConstraints['TCustomId'] = string,
-  in out TId extends RouteConstraints['TId'] = ResolveId<
-    TParentRoute,
+    TPath,
+    TFullPath,
     TCustomId,
-    TPath
-  >,
-  in out TSearchValidator = undefined,
-  in out TParams = ResolveParams<TPath>,
-  in out TRouterContext = AnyContext,
-  in out TRouteContextFn = AnyContext,
-  in out TBeforeLoadFn = AnyContext,
-  in out TLoaderDeps extends Record<string, any> = {},
-  in out TLoaderFn = undefined,
-  in out TChildren = unknown,
-  in out TFileRouteTypes = unknown,
-> extends BaseRoute<
-  TParentRoute,
-  TPath,
-  TFullPath,
-  TCustomId,
-  TId,
-  TSearchValidator,
-  TParams,
-  TRouterContext,
-  TRouteContextFn,
-  TBeforeLoadFn,
-  TLoaderDeps,
-  TLoaderFn,
-  TChildren,
-  TFileRouteTypes
-> {
+    TId,
+    TSearchValidator,
+    TParams,
+    TRouterContext,
+    TRouteContextFn,
+    TBeforeLoadFn,
+    TLoaderDeps,
+    TLoaderFn,
+    TChildren,
+    TFileRouteTypes
+  >
+  implements
+    RouteCore<
+      TParentRoute,
+      TPath,
+      TFullPath,
+      TCustomId,
+      TId,
+      TSearchValidator,
+      TParams,
+      TRouterContext,
+      TRouteContextFn,
+      TBeforeLoadFn,
+      TLoaderDeps,
+      TLoaderFn,
+      TChildren,
+      TFileRouteTypes
+    >
+{
   /**
    * @deprecated Use the `createRoute` function instead.
    */
@@ -364,24 +385,37 @@ export function createRootRouteWithContext<TRouterContext extends {}>() {
 export const rootRouteWithContext = createRootRouteWithContext
 
 export class RootRoute<
-  in out TSearchValidator = undefined,
-  in out TRouterContext = {},
-  in out TRouteContextFn = AnyContext,
-  in out TBeforeLoadFn = AnyContext,
-  in out TLoaderDeps extends Record<string, any> = {},
-  in out TLoaderFn = undefined,
-  in out TChildren = unknown,
-  in out TFileRouteTypes = unknown,
-> extends BaseRootRoute<
-  TSearchValidator,
-  TRouterContext,
-  TRouteContextFn,
-  TBeforeLoadFn,
-  TLoaderDeps,
-  TLoaderFn,
-  TChildren,
-  TFileRouteTypes
-> {
+    in out TSearchValidator = undefined,
+    in out TRouterContext = {},
+    in out TRouteContextFn = AnyContext,
+    in out TBeforeLoadFn = AnyContext,
+    in out TLoaderDeps extends Record<string, any> = {},
+    in out TLoaderFn = undefined,
+    in out TChildren = unknown,
+    in out TFileRouteTypes = unknown,
+  >
+  extends BaseRootRoute<
+    TSearchValidator,
+    TRouterContext,
+    TRouteContextFn,
+    TBeforeLoadFn,
+    TLoaderDeps,
+    TLoaderFn,
+    TChildren,
+    TFileRouteTypes
+  >
+  implements
+    RootRouteCore<
+      TSearchValidator,
+      TRouterContext,
+      TRouteContextFn,
+      TBeforeLoadFn,
+      TLoaderDeps,
+      TLoaderFn,
+      TChildren,
+      TFileRouteTypes
+    >
+{
   /**
    * @deprecated `RootRoute` is now an internal implementation detail. Use `createRootRoute()` instead.
    */

--- a/packages/router-core/src/index.ts
+++ b/packages/router-core/src/index.ts
@@ -198,6 +198,7 @@ export type {
   RouteAddChildrenFn,
   RouteAddFileChildrenFn,
   RouteAddFileTypesFn,
+  RootRoute,
 } from './route'
 
 export {

--- a/packages/router-core/src/route.ts
+++ b/packages/router-core/src/route.ts
@@ -620,7 +620,24 @@ export interface Route<
       TBeforeLoadFn
     >,
   ) => this
-  lazy: RouteLazyFn<this>
+  lazy: RouteLazyFn<
+    Route<
+      TParentRoute,
+      TPath,
+      TFullPath,
+      TCustomId,
+      TId,
+      TSearchValidator,
+      TParams,
+      TRouterContext,
+      TRouteContextFn,
+      TBeforeLoadFn,
+      TLoaderDeps,
+      TLoaderFn,
+      TChildren,
+      TFileRouteTypes
+    >
+  >
   addChildren: RouteAddChildrenFn<
     TParentRoute,
     TPath,
@@ -1269,24 +1286,7 @@ export class BaseRoute<
   in out TLoaderFn = undefined,
   in out TChildren = unknown,
   in out TFileRouteTypes = unknown,
-> implements
-    Route<
-      TParentRoute,
-      TPath,
-      TFullPath,
-      TCustomId,
-      TId,
-      TSearchValidator,
-      TParams,
-      TRouterContext,
-      TRouteContextFn,
-      TBeforeLoadFn,
-      TLoaderDeps,
-      TLoaderFn,
-      TChildren,
-      TFileRouteTypes
-    >
-{
+> {
   isRoot: TParentRoute extends AnyRoute ? true : false
   options: RouteOptions<
     TParentRoute,
@@ -1561,7 +1561,24 @@ export class BaseRoute<
     return this
   }
 
-  lazy: RouteLazyFn<this> = (lazyFn) => {
+  lazy: RouteLazyFn<
+    Route<
+      TParentRoute,
+      TPath,
+      TFullPath,
+      TCustomId,
+      TId,
+      TSearchValidator,
+      TParams,
+      TRouterContext,
+      TRouteContextFn,
+      TBeforeLoadFn,
+      TLoaderDeps,
+      TLoaderFn,
+      TChildren,
+      TFileRouteTypes
+    >
+  > = (lazyFn) => {
     this.lazyFn = lazyFn
     return this
   }
@@ -1578,6 +1595,32 @@ export class BaseRouteApi<TId, TRouter extends AnyRouter = RegisteredRouter> {
     return notFound({ routeId: this.id as string, ...opts })
   }
 }
+
+export interface RootRoute<
+  in out TSearchValidator = undefined,
+  in out TRouterContext = {},
+  in out TRouteContextFn = AnyContext,
+  in out TBeforeLoadFn = AnyContext,
+  in out TLoaderDeps extends Record<string, any> = {},
+  in out TLoaderFn = undefined,
+  in out TChildren = unknown,
+  in out TFileRouteTypes = unknown,
+> extends Route<
+    any, // TParentRoute
+    '/', // TPath
+    '/', // TFullPath
+    string, // TCustomId
+    RootRouteId, // TId
+    TSearchValidator, // TSearchValidator
+    {}, // TParams
+    TRouterContext,
+    TRouteContextFn,
+    TBeforeLoadFn,
+    TLoaderDeps,
+    TLoaderFn,
+    TChildren, // TChildren
+    TFileRouteTypes
+  > {}
 
 export class BaseRootRoute<
   in out TSearchValidator = undefined,

--- a/packages/solid-router/src/route.tsx
+++ b/packages/solid-router/src/route.tsx
@@ -24,9 +24,11 @@ import type {
   ResolveFullPath,
   ResolveId,
   ResolveParams,
+  RootRoute as RootRouteCore,
   RootRouteId,
   RootRouteOptions,
   RouteConstraints,
+  Route as RouteCore,
   RouteIds,
   RouteMask,
   RouteOptions,
@@ -142,43 +144,62 @@ export class RouteApi<
 }
 
 export class Route<
-  in out TParentRoute extends RouteConstraints['TParentRoute'] = AnyRoute,
-  in out TPath extends RouteConstraints['TPath'] = '/',
-  in out TFullPath extends RouteConstraints['TFullPath'] = ResolveFullPath<
+    in out TParentRoute extends RouteConstraints['TParentRoute'] = AnyRoute,
+    in out TPath extends RouteConstraints['TPath'] = '/',
+    in out TFullPath extends RouteConstraints['TFullPath'] = ResolveFullPath<
+      TParentRoute,
+      TPath
+    >,
+    in out TCustomId extends RouteConstraints['TCustomId'] = string,
+    in out TId extends RouteConstraints['TId'] = ResolveId<
+      TParentRoute,
+      TCustomId,
+      TPath
+    >,
+    in out TSearchValidator = undefined,
+    in out TParams = ResolveParams<TPath>,
+    in out TRouterContext = AnyContext,
+    in out TRouteContextFn = AnyContext,
+    in out TBeforeLoadFn = AnyContext,
+    in out TLoaderDeps extends Record<string, any> = {},
+    in out TLoaderFn = undefined,
+    in out TChildren = unknown,
+    in out TFileRouteTypes = unknown,
+  >
+  extends BaseRoute<
     TParentRoute,
-    TPath
-  >,
-  in out TCustomId extends RouteConstraints['TCustomId'] = string,
-  in out TId extends RouteConstraints['TId'] = ResolveId<
-    TParentRoute,
+    TPath,
+    TFullPath,
     TCustomId,
-    TPath
-  >,
-  in out TSearchValidator = undefined,
-  in out TParams = ResolveParams<TPath>,
-  in out TRouterContext = AnyContext,
-  in out TRouteContextFn = AnyContext,
-  in out TBeforeLoadFn = AnyContext,
-  in out TLoaderDeps extends Record<string, any> = {},
-  in out TLoaderFn = undefined,
-  in out TChildren = unknown,
-  in out TFileRouteTypes = unknown,
-> extends BaseRoute<
-  TParentRoute,
-  TPath,
-  TFullPath,
-  TCustomId,
-  TId,
-  TSearchValidator,
-  TParams,
-  TRouterContext,
-  TRouteContextFn,
-  TBeforeLoadFn,
-  TLoaderDeps,
-  TLoaderFn,
-  TChildren,
-  TFileRouteTypes
-> {
+    TId,
+    TSearchValidator,
+    TParams,
+    TRouterContext,
+    TRouteContextFn,
+    TBeforeLoadFn,
+    TLoaderDeps,
+    TLoaderFn,
+    TChildren,
+    TFileRouteTypes
+  >
+  implements
+    RouteCore<
+      TParentRoute,
+      TPath,
+      TFullPath,
+      TCustomId,
+      TId,
+      TSearchValidator,
+      TParams,
+      TRouterContext,
+      TRouteContextFn,
+      TBeforeLoadFn,
+      TLoaderDeps,
+      TLoaderFn,
+      TChildren,
+      TFileRouteTypes
+    >
+{
   /**
    * @deprecated Use the `createRoute` function instead.
    */
@@ -352,24 +373,37 @@ export function createRootRouteWithContext<TRouterContext extends {}>() {
 export const rootRouteWithContext = createRootRouteWithContext
 
 export class RootRoute<
-  in out TSearchValidator = undefined,
-  in out TRouterContext = {},
-  in out TRouteContextFn = AnyContext,
-  in out TBeforeLoadFn = AnyContext,
-  in out TLoaderDeps extends Record<string, any> = {},
-  in out TLoaderFn = undefined,
-  in out TChildren = unknown,
-  in out TFileRouteTypes = unknown,
-> extends BaseRootRoute<
-  TSearchValidator,
-  TRouterContext,
-  TRouteContextFn,
-  TBeforeLoadFn,
-  TLoaderDeps,
-  TLoaderFn,
-  TChildren,
-  TFileRouteTypes
-> {
+    in out TSearchValidator = undefined,
+    in out TRouterContext = {},
+    in out TRouteContextFn = AnyContext,
+    in out TBeforeLoadFn = AnyContext,
+    in out TLoaderDeps extends Record<string, any> = {},
+    in out TLoaderFn = undefined,
+    in out TChildren = unknown,
+    in out TFileRouteTypes = unknown,
+  >
+  extends BaseRootRoute<
+    TSearchValidator,
+    TRouterContext,
+    TRouteContextFn,
+    TBeforeLoadFn,
+    TLoaderDeps,
+    TLoaderFn,
+    TChildren,
+    TFileRouteTypes
+  >
+  implements
+    RootRouteCore<
+      TSearchValidator,
+      TRouterContext,
+      TRouteContextFn,
+      TBeforeLoadFn,
+      TLoaderDeps,
+      TLoaderFn,
+      TChildren,
+      TFileRouteTypes
+    >
+{
   /**
    * @deprecated `RootRoute` is now an internal implementation detail. Use `createRootRoute()` instead.
    */

